sdl-core/src/main/scala/io/smartdatalake/util/spark/SDLSparkExtension.scala

```diff
@@ -0,0 +1,78 @@
+/*
+ * Smart Data Lake - Build your data lake the smart way.
+ *
+ * Copyright © 2019-2024 ELCA Informatique SA (<https://www.elca.ch>)
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program. If not, see <http://www.gnu.org/licenses/>.
+ */
+
+package io.smartdatalake.util.spark
+
+import io.smartdatalake.definitions.Environment
+import org.apache.spark.sql.catalyst.expressions.Attribute
+import org.apache.spark.sql.catalyst.plans.logical.{LogicalPlan, UnaryNode}
+import org.apache.spark.sql.execution.{SparkPlan, UnaryExecNode}
+import org.apache.spark.sql.{DataFrame, DatasetUtils, SparkSessionExtensions, Strategy}
+
+/**
+ * Custom extensions for the Spark Session, e.g. optimizer rules or additional plan strategies.
+ */
+private[smartdatalake] class SDLSparkExtension extends (SparkSessionExtensions => Unit) {
+  def apply(e: SparkSessionExtensions): Unit = {
+    e.injectPlannerStrategy((parameters) => AssertNotEmptyStrategy)
+  }
+}
+
+object SDLSparkExtension {
+  /**
+   * Fail on evaluation of DataFrame if it is empty.
+   */
+  def assertNotEmpty(df: DataFrame): DataFrame = {
+    // TODO: use DatasetHelper from spark-expressions instead
+    if (Environment.enableSparkPlanAssertNotEmpty) DatasetUtils.applyOperator(df, plan => AssertNotEmpty(plan))
+    else df
+  }
+}
+
+
+object AssertNotEmptyStrategy extends Strategy {
+  def apply(plan: LogicalPlan) = plan match {
+    case AssertNotEmpty(child) =>
+      AssertNotEmptyExec(planLater(child)) :: Nil
+    case _ => Nil
+  }
+}
+
+private[smartdatalake] case class AssertNotEmpty(child: LogicalPlan) extends UnaryNode {
+  override def output: Seq[Attribute] = child.output
+  override protected def withNewChildInternal(newChild: LogicalPlan): LogicalPlan = copy(child = newChild)
+}
+
+case class AssertNotEmptyExec(child: SparkPlan) extends UnaryExecNode {
+
+  // Output has the same attributes as input
+  override def output: Seq[Attribute] = child.output
+
+  // Examine child's data
+  override protected def doExecute() = {
+    val rdd = child.execute()
```

Comment:
is Spark smart enough to not completely execute the whole query plan a second time for this? It does not look like it

Reply:
Thanks, you're right. This triggers a separate spark job (within the same Query), which is not optimal.
I now found different ways to evaluate the logical plan, without using rdd count or isEmpty.

Reply:
I don t get what you changed compared to before, let's have a look together